chain-signatures/node/src/protocol/triple.rs

```diff
@@ -185,56 +192,18 @@ impl TripleManager {
         id0: TripleId,
         id1: TripleId,
     ) -> Result<(Triple, Triple), GenerationError> {
-        let triples = &self.triple_storage;
-        let triple_0 = match triples.take(&id0).await {
-            Ok(Some(triple)) => triple,
-            Ok(None) => {
-                if self.generators.contains_key(&id0) {
-                    tracing::warn!(id0, "triple is generating");
-                    return Err(GenerationError::TripleIsGenerating(id0));
-                } else if self.gc.contains_key(&id0) {
-                    tracing::warn!(id0, "triple is garbage collected");
-                    return Err(GenerationError::TripleIsGarbageCollected(id0));
-                } else {
-                    tracing::warn!(id0, "triple is missing");
-                    return Err(GenerationError::TripleIsMissing(id0));
-                }
-            }
-            Err(e) => {
-                tracing::warn!(id0, ?e, "failed to take triple");
-                return Err(GenerationError::TripleIsMissing(id0));
-            }
-        };
-
-        let triple_1 = match triples.take(&id1).await {
-            Ok(Some(triple)) => triple,
-            Ok(None) => {
-                if let Err(e) = triples.insert(triple_0).await {
-                    tracing::warn!(id0, ?e, "failed to insert triple back");
+        let triple_0 = self.take(&id0).await?;
+        let triple_1 = match self.take(&id1).await {
+            Ok(triple) => triple,
+            Err(err) => {
+                if let Err(store_err) = self.triple_storage.insert(triple_0, false).await {
+                    tracing::warn!(?store_err, id0, "failed to insert triple back");
                 }
-                if self.generators.contains_key(&id1) {
-                    tracing::warn!(id1, "triple is generating");
-                    return Err(GenerationError::TripleIsGenerating(id1));
-                } else if self.gc.contains_key(&id1) {
-                    tracing::warn!(id1, "triple is garbage collected");
-                    return Err(GenerationError::TripleIsGarbageCollected(id1));
-                } else {
-                    tracing::warn!(id1, "triple is missing");
-                    return Err(GenerationError::TripleIsMissing(id1));
-                }
-            }
-            Err(e) => {
-                tracing::warn!(id1, ?e, "failed to take triple");
-                if let Err(e) = triples.insert(triple_0).await {
-                    tracing::warn!(id0, ?e, "failed to insert triple back");
-                }
-                return Err(GenerationError::TripleIsMissing(id1));
+                return Err(err);
             }
         };
-
         self.gc.insert(id0, Instant::now());
```

Comment:
Let's not do gc.insert here, let's do it in fn take() so that it's all done in take(), and when the 2nd triple take fails in take_two(), the insert() will revert the gc.insert by a gc.remove

Reply:
this is still unchanged tho

Reply:
I wonder if you still had an older view of the code, but I changed it already